{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import tables\n",
    "import io\n",
    "import scipy\n",
    "import numpy\n",
    "from scipy import sparse\n",
    "import pandas\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "HOST_URL = \"https://scrap.caltech.edu/\"\n",
    "SEQUENCING_RUN_NAME = \"JPMT35\"\n",
    "\n",
    "SAMPLES = [\n",
    "    \"PBMC-MULT-14-1\",\n",
    "    \"PBMC-MULT-14-2\",\n",
    "    \"PBMC-MULT-14-3\",\n",
    "    \"PBMC-MULT-14-4\",\n",
    "    \"PBMC-MULT-14-5\",\n",
    "    \"PBMC-MULT-14-6\",\n",
    "    \"PBMC-MULT-16-1\",\n",
    "    \"PBMC-MULT-16-2\",\n",
    "    \"PBMC-MULT-16-3\",\n",
    "    \"PBMC-MULT-16-4\",\n",
    "    \"PBMC-MULT-16-5\",\n",
    "    \"PBMC-MULT-16-6\",\n",
    "    \"PBMC-MULT-14-1-TAGS\",\n",
    "    \"PBMC-MULT-14-2-TAGS\",\n",
    "    \"PBMC-MULT-14-3-TAGS\",\n",
    "    \"PBMC-MULT-14-4-TAGS\",\n",
    "    \"PBMC-MULT-14-5-TAGS\",\n",
    "    \"PBMC-MULT-14-6-TAGS\",\n",
    "    \"PBMC-MULT-16-1-TAGS\",\n",
    "    \"PBMC-MULT-16-2-TAGS\",\n",
    "    \"PBMC-MULT-16-3-TAGS\",\n",
    "    \"PBMC-MULT-16-4-TAGS\",\n",
    "    \"PBMC-MULT-16-5-TAGS\",\n",
    "    \"PBMC-MULT-16-6-TAGS\"\n",
    "]\n",
    "\n",
    "# CELL_SET_NAMES = [\n",
    "#     \"LC2-B-V-2\",\n",
    "#     \"LM1-B-V\",\n",
    "#     \"LM1-S-2\"\n",
    "# ]\n",
    "\n",
    "CELL_SET_NAMES = SAMPLES\n",
    "\n",
    "SAMPLE_FASTQ_PREFIXES = SAMPLES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a cell set for each sample, or get it if it already exists\n",
    "cell_sets_URL = HOST_URL + \"cell_sets\"\n",
    "\n",
    "response = requests.get(cell_sets_URL, verify=False)\n",
    "cell_sets = response.json()[\"cell_sets\"]\n",
    "\n",
    "sample_cell_sets = {}\n",
    "\n",
    "for sample_index, sample in enumerate(SAMPLES):\n",
    "    cell_set_id = None\n",
    "\n",
    "    for cell_set in cell_sets:\n",
    "        if cell_set[\"name\"] == CELL_SET_NAMES[sample_index]:\n",
    "            cell_set_id = cell_set[\"_id\"]\n",
    "            break\n",
    "\n",
    "    if not cell_set_id:\n",
    "        cell_set = {}\n",
    "        cell_set[\"name\"] = sample\n",
    "        \n",
    "        print(\"Creating Cell Set %s\" % sample)\n",
    "        \n",
    "        response = requests.post(cell_sets_URL, json=cell_set, verify=False)\n",
    "        cell_set = response.json()\n",
    "        sample_cell_sets[sample] = cell_set\n",
    "    else:\n",
    "        sample_cell_sets[sample] = cell_set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find the sequencing run associated with these samples, or create it if it doesn't exist\n",
    "sequencing_runs_URL = HOST_URL + \"sequencing_runs\"\n",
    "\n",
    "response = requests.get(sequencing_runs_URL, verify=False)\n",
    "sequencing_runs = response.json()[\"sequencing_runs\"]\n",
    "\n",
    "sequencing_run_id = None\n",
    "\n",
    "for sequencing_run in sequencing_runs:\n",
    "    if sequencing_run[\"name\"] == SEQUENCING_RUN_NAME:\n",
    "        sequencing_run_id = sequencing_run[\"_id\"]\n",
    "        break\n",
    "\n",
    "if not sequencing_run_id:\n",
    "    sequencing_run = {}\n",
    "    sequencing_run[\"type\"] = \"local_BCLs\"\n",
    "    sequencing_run[\"name\"] = SEQUENCING_RUN_NAME\n",
    "    sequencing_run[\"status\"] = \"available\"\n",
    "    \n",
    "    print(\"Creating sequencing run %s\" % SEQUENCING_RUN_NAME)\n",
    "    response = requests.post(sequencing_runs_URL, json=sequencing_run, verify=False)\n",
    "    sequencing_run_id = response.json()[\"_id\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sequencing_run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "S3_SEARCH_DIRECTORY = \"sequencing_runs/20200116T013344Z_JPMT35-2/200109_A00351_0310_AH2MH3DSXY_10x/JPMT35/\"\n",
    "\n",
    "read_sets_URL = HOST_URL + \"read_sets\"\n",
    "response = requests.get(read_sets_URL, verify=False)\n",
    "read_sets = response.json()[\"read_sets\"]\n",
    "\n",
    "out = os.popen('aws s3 ls s3://scrap-dm/%s' % S3_SEARCH_DIRECTORY).read() # list all files found in an S3 folder\n",
    "out = out.split('\\n') # create a list from string output\n",
    "out = [x.split(' ')[-1] for x in out if x.endswith('gz')] # only get .gz elements and file names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for sample_index, sample_name in enumerate(SAMPLES):\n",
    "    \n",
    "    read_set_id = None\n",
    "    \n",
    "    for read_set in read_sets:\n",
    "        if read_set[\"name\"] == sample_name:\n",
    "            read_set_id = read_set[\"_id\"]\n",
    "            break\n",
    "\n",
    "    if not read_set_id:\n",
    "        read_set = {}\n",
    "        read_set[\"name\"] = sample_name\n",
    "        read_set[\"status\"] = \"available\"\n",
    "        \n",
    "        print(\"Creating read set %s\" % sample_name)\n",
    "        \n",
    "        response = requests.post(read_sets_URL, json=cell_set, verify=False)\n",
    "        read_set = response.json()\n",
    "        read_set[\"sequencing_run_id\"] = sequencing_run_id\n",
    "        read_set[\"status\"] = \"available\"\n",
    "        read_set[\"cell_set_id\"] = sample_cell_sets[sample_name][\"_id\"]\n",
    "        \n",
    "        response = requests.put(read_sets_URL + \"/%s\" % read_set[\"_id\"], json=read_set, verify=False)\n",
    "        read_set = response.json()\n",
    "    \n",
    "    print(read_set[\"path_id\"])\n",
    "    \n",
    "    destination_files = os.popen('aws s3 ls s3://scrap-dm/reads/%s/' % read_set[\"path_id\"]).read() # list all files found in an S3 folder\n",
    "    destination_files = destination_files.split('\\n') # create a list from string output\n",
    "    destination_files = [x.split(' ')[-1] for x in destination_files if x.endswith('.fastq.gz')] # only get .gz elements and file names\n",
    "    \n",
    "    source_files = []\n",
    "    \n",
    "    for file_name in out:\n",
    "        file_sample_name = file_name.split(\"_\")\n",
    "        file_sample_name = \"_\".join(file_sample_name[0:-4])\n",
    "        if file_sample_name == SAMPLE_FASTQ_PREFIXES[sample_index]:\n",
    "            source_files.append(file_name)\n",
    "            \n",
    "    print(\"Source files:\")\n",
    "    print(source_files)\n",
    "    print(\"Destination files:\")\n",
    "    print(destination_files)\n",
    "    \n",
    "    for source_file in source_files:\n",
    "        new_file_name = source_file\n",
    "        while new_file_name in destination_files:\n",
    "            new_file_name_parts = new_file_name[0:-9].split(\"_\")\n",
    "            new_file_name = \"_\".join(new_file_name_parts[0:-1]) + \"_%03d\" % (int(new_file_name_parts[-1])+1) + \".fastq.gz\"\n",
    "        \n",
    "        if new_file_name != source_file:\n",
    "            print(\"Will rename %s to %s\" % (source_file, new_file_name))\n",
    "            \n",
    "        source_path = \"s3://scrap-dm/%s%s\" % (S3_SEARCH_DIRECTORY, source_file)\n",
    "        destination_path = \"s3://scrap-dm/reads/%s/%s\" % (read_set[\"path_id\"], new_file_name)\n",
    "        print(\"Moving %s to %s\" % (source_path, destination_path))\n",
    "        os.system('aws s3 mv %s %s' % (source_path, destination_path))\n",
    "        read_set[\"FASTQ_files\"][new_file_name] = {\"remote_path\" : new_file_name, \"status\" : \"available\"}\n",
    "        response = requests.put(read_sets_URL + \"/%s\" % read_set[\"_id\"], json=read_set, verify=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
